### ISSUE TITLE
use lossless serialization function for kv-store

### ISSUE THREAD

**Issue by BrunoBonacci** [samsara/samsara#46]
the current solution to store state during the processing is by sending the key/value updates
to a special (compacted) topic.
The format currently utilized is json. Unfortunately json data types are very poor compared to EDNs.
So if you store a set (`#{}`) of even a keyword in place of a value, they lose their type after the first round-trip (store/restore).

We need to be able to store and load data without losing the semantic characteristic of certain types, or at least have multiple possibility which could be selected depending of the case.

There are a number of interesting solution for storing and loading the data which we should support.
Here is a few:

  - [nippy](https://github.com/ptaoussanis/nippy)
  - [fressian](https://github.com/clojure/data.fressian)
  - [json](https://github.com/dakrone/cheshire)
  - EDN Clojure
  - [transit](https://github.com/cognitect/transit-clj)

Clone Samsara and complete the library called `scribe` under `/internal/scribe` folder.
Create a multimethod function called `scribe` which take  a configuration and return 
a instance of the `Scribe` protocol.

`(scribe {:type :nippy :more :config})`

The Scribe protocol has two functions: 

     (write this data) -> bytes 
     (read this bytes) -> data

Then implement implementation of the Scribe protocol with the format listed above.

**Comment by puhrez** [comment]
nippy seems like the fastest, most active, and most documented of the two.

**Comment by BrunoBonacci** [comment]
I've done a few tests/benchmarks on possible solutions to use in this space https://github.com/BrunoBonacci/fast-data

The problem is that we want to be able to use in-memory clojure data structure for our processing for higher 0performance.
The in-memory data is then backed into a persistent store (topic/db). So we need to find a serialization and deserialization mechanism to store the data which it won't lose the original clojure type when serialized and the deserialized.

In my tests, all tool i've tried they all have problem in the round-trip, but overall nippy seems to be the best choice (performance/completeness).

anyway we will have to be able to support multiple options so we will have to put this behind a protocol/factory.

@puhrez: If you wish to contribute this i can fill you up with the other details

**Comment by puhrez** [comment]
@BrunoBonacci Yea I'm interested! right now I'm exploring your `fast-data` repo and running it.

**Comment by BrunoBonacci** [comment]
`transit` was an interesting choice because it is also (almost) human readable,
However the authors discourage the storing of transit data. As the name implies it should be used only for data in transit and not at rest.

from: https://github.com/cognitect/transit-clj
> NOTE: Transit is a work in progress and may evolve based on feedback. As a result, while Transit is a great option for transferring data between applications, **it should not yet be used for storing data durably over time**. This recommendation will change when the specification is complete.

So I think the best is to default to nippy but put it configurable, what's your thoughts?

**Comment by puhrez** [comment]
hmmm, given that the last update to the `transit`'s src was 11 months ago, I'm unsure whether the specification will ever be complete...

Moreover if we peg it a specific version we should be safe from future work no?

regardless, nippy does seem to be the best default given its performance with user-data which I think is rather representative of data being processed (strings, numericals, boolean, lists).

that being said, it might be best to do more specific type tests, because even edn destroys nippy for the basic data test which covers so many types...


TL;DR yes configurable, with nippy as a default.

**Comment by BrunoBonacci** [comment]
> hmmm, given that the last update to the transit's src was 11 months ago, I'm unsure whether the specification will ever be complete...

Actually I would say that this is a good sign that the format is now stabilizing.

> Moreover if we peg it a specific version we should be safe from future work no?

That's not a good strategy. what it they fix a bug? we will want to get it.

Ok, I will add draft of design and what I think should be done, so that we can discuss it and take it from there.

**Comment by puhrez** [comment]
what I meant to say, is that given that the last update was 11 months ago and they say that it the specification is not complete, that you're right, its probably stabilized.

> That's not a good strategy. what it they fix a bug? we will want to get it.

true

**Comment by kolotaev** [comment]
Implemented here: https://github.com/samsara/samsara/pull/68

**Comment by BrunoBonacci** [comment]
fixed with #68
